internal/cmdrpkgfindupdates/command.go

```diff
@@ -0,0 +1,221 @@
+	r := &runner{
+		ctx:    ctx,
+		cfg:    rcg,
+		client: nil,
+	}
+
+	c := &cobra.Command{
+		Use: "find-updates PACKAGE_REVISION",
+		//Short:   rpkgdocs.FindUpdatesShort,
+		//Long:    rpkgdocs.FindUpdatesShort + "\n" + rpkgdocs.FindUpdatesLong,
+		//Example: rpkgdocs.FindUpdatesExamples,
+		PreRunE: r.preRunE,
+		RunE:    r.runE,
+		Hidden:  porch.HidePorchCommands,
+	}
+	r.Command = c
+	return r
+}
+
+type runner struct {
+	ctx     context.Context
+	cfg     *genericclioptions.ConfigFlags
+	client  client.Client
+	Command *cobra.Command
+
+	// Flags
+}
+
+func (r *runner) preRunE(_ *cobra.Command, args []string) error {
+	const op errors.Op = command + ".preRunE"
+
+	if len(args) < 1 {
+		return errors.E(op, "PACKAGE_REVISION is a required positional argument")
+	}
+	cf, err := r.cfg.ToRESTConfig()
+	if err != nil {
+		return errors.E(op, err)
+	}
+	scheme, err := rpkgutil.CreateScheme()
+	if err != nil {
+		return errors.E(op, err)
+	}
+	c, err := client.New(cf, client.Options{Scheme: scheme})
+	if err != nil {
+		return errors.E(op, err)
+	}
+	r.client = c
+	return nil
+}
+
+func (r *runner) runE(cmd *cobra.Command, args []string) error {
+	const op errors.Op = command + ".runE"
+	var messages []string
+
+	for _, p := range args {
+		packageRevision, err := r.getPackageRevision(p)
+		if err != nil {
+			return errors.E(op, err)
+		}
+
+		upstreamLock, err := r.getUpstreamLock(packageRevision)
+		if err != nil {
+			return errors.E(op, err)
+		}
+
+		availableUpdates, err := r.availableUpdates(upstreamLock)
+		if err != nil {
+			return errors.E(op, err)
+		}
+
+		fmt.Fprintf(cmd.OutOrStdout(), "%s %s\n", p, availableUpdates)
+	}
+
+	if len(messages) > 0 {
+		return errors.E(op, fmt.Errorf("errors:\n  %s", strings.Join(messages, "\n  ")))
+	}
+
+	return nil
+}
+
+func (r *runner) getPackageRevision(p string) (porchapi.PackageRevision, error) {
+	packageRevision := porchapi.PackageRevision{}
+	newScheme := runtime.NewScheme()
+	if err := porchapi.SchemeBuilder.AddToScheme(newScheme); err != nil {
+		return packageRevision, err
+	}
+	restClient, err := porch.CreateRESTClient(r.cfg)
+	if err != nil {
+		return packageRevision, err
+	}
+	err = restClient.
+		Get().
+		Namespace(*r.cfg.Namespace).
+		Resource("packagerevisions").
+		Name(p).
+		Do(r.ctx).
+		Into(&packageRevision)
+
+	return packageRevision, err
+}
+
+func (r *runner) getUpstreamLock(packageRevision porchapi.PackageRevision) (*kptfile.UpstreamLock, error) {
+	var resources porchapi.PackageRevisionResources
+	if err := r.client.Get(r.ctx, client.ObjectKey{
+		Namespace: *r.cfg.Namespace,
+		Name:      packageRevision.Name,
+	}, &resources); err != nil {
+		return nil, err
+	}
+
+	// get the upstream package URL from the Kptfile
+	var kf *kptfile.KptFile
+	for k, v := range resources.Spec.Resources {
+		if k == kptfile.KptFileName {
+			var err error
+			kf, err = pkg.DecodeKptfile(strings.NewReader(v))
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
```

Comment:
This was discussed briefly in yesterday's meeting.

porch API should expose `upstream` and `upstreamLock` info. `UpstreamLock` aligns pretty well to be exposed in the status field of a packagerevision. `upstream` is little bit tricky because of tasks, but having it as annotation seems like a good starting point to me.

Reply:
SG. I can try to integrate that into this PR.

FYI @mortent

Reply:
actually, do we need to expose `upstream` in the porch API? ? It seems that `upstreamLock` to me is enough for now

Reply:
Yeah, I think you are correct that we don't need to expose the `upstream` for now. The `UpstreamLock` is also already available through the internal `PackageRevision` interface: https://github.com/GoogleContainerTools/kpt/blob/059babc3d801841d25b63950dfcfdf261f236600/porch/pkg/repository/repository.go#L61

Reply:
The git implementation of `GetUpstreamLock` seems to not be working correctly. I'll work on that in another PR.

Reply:
https://github.com/GoogleContainerTools/kpt/pull/3301